### PR TITLE
Support register_hook on input tensors

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107502
* __->__ #108903
* #109092
* #108846

The strategy in this PR is pretty straightforward.

There are 2 kinds of hooks:

1) Hooks on objects with sources (inputs, params)
2) Hooks on objects w/o sources (intermediaries, and outputs). 

Note: As outputs can be made simple by how dynamo handles residuals, they could actually be handled as if they were inputs, but, for the sake of this PR, we will refer to hooks as either hooks on inputs (sourced), or hooks on intermediaries (not sourced).

The plan:

**For tensors w/ a source:**
We record registered hooks, store them as a global, and associate them with the tensor in residuals. This means that when dynamo goes to create the frame, where we produce bytecode to stitch together our PT2 modified bytecode with the original eager code, we call `register_hook`. This registration of hooks in residuals is sound because (a) it happens right after a Pt2 frame region ends and (b) we know that the tensor is alive in f_locals, f_globals, or a module in the users invoking frame. This means we can soundly know it will be around to invoke `register_hook` on. As long as we guard on the identity of the lifted function, this is sound to do. 

**For tensors w/o a source:**
Graph break - we will support this in a subsequent PR 

**Handles:**

An interesting new component here is the creation of a `STORE_FAST `->`LOAD_FAST` associated with the handle, the return result of `register_hook`. If the user code stored the result of `register_hook` in a handle, we need to honor that. We do so by interceding into `STORE_FAST`, and recording the name of the local variable as directed by user code. We then honor that same name in the reconstructed bytecode. If the user did not store a hook, we merely pop the produced value to preserve the stack.

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: test/dynamo/test_hooks.py [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+        v = torch.tensor([0.0, 0.0, 0.0], requires_grad=True)
+        v = fn(v, torch.randn([2, 2]), torch.randn([2, 2]))[0]
+        v.backward(torch.tensor([1.0, 2.0, 3.0]))
+        self.assertEqual(v.grad, torch.tensor([3.0, 6.0, 9.0]))
+        self.assertEqual(cnts.frame_count, 2)
+
+    def test_tensor_register_hook_multi_handle_return(self):
+        def fn(x, y, z):
+            handle = x.register_hook(lambda grad: grad * 2)
+            h2 = handle
```
Comment (voznesenskym) [inline test/dynamo/test_hooks.py:74]: @ezyang this is one case you had questions about / covered w/r/t name preservation.

File: test/dynamo/test_hooks.py [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+        v.backward(torch.tensor([1.0, 2.0, 3.0]))
+        self.assertEqual(v.grad, torch.tensor([2.0, 4.0, 6.0]))
+        self.assertEqual(cnts.frame_count, 1)
+        self.assertNotEqual(h, None)
+        self.assertNotEqual(h2, None)
+        self.assertEqual(h2, h)
+
+    def test_tensor_register_hook_repeated_handle_not_local(self):
+        def fn(x, y, z, mod):
+            mod.handle = x.register_hook(lambda grad: grad * 2)
```
Comment (voznesenskym) [inline test/dynamo/test_hooks.py:109]: @ezyang this is another case you had questions about / covered w/r/t name preservation.

File: torch/_dynamo/side_effects.py [hunk 3]
```diff
@@ -375,6 +380,56 @@ def codegen_save_tempvars(self, cg: PyCodegen):
...
+    def codegen_hooks(self, cg):
+        for (
+            tensor,
+            hook,
+            handle,
+        ) in self.tensor_hooks:
+            # Note: [On tensor.register_hook]
+            #
+            # register_hook on a tensor, AKA backward hooks, have slightly nuanced differences in how they are implemented
+            # when it comes to hooks on objects with sources (inputs, params) vs objects without sources (intermediaries).
```
Comment (bdhirsh) [inline torch/_dynamo/side_effects.py:395]: Is it worth explaining why these two cases need to be differentiated?

My understanding (lmk if this sounds wrong) is: hooks on intermediates and outputs can be captured directly by AOTAutograd, since they get sandwiched inside of the backward graph when AOTAutograd calls `autograd.grad(graph_outputs, graph_inputs)`.

But hooks on intermediates aren't supposed to run until **after** we've computed gradients for the graph inputs, which means that they need to run outside of AOTAutograd's backward graph. So instead, we rely on compiled autograd to first trace through AOTAutograd's captured backward graph, and then trace through any hooks on the grad_inputs (which dynamo installs in the residuals, here).

File: torch/_dynamo/side_effects.py [hunk 4]
```diff
@@ -375,6 +380,56 @@ def codegen_save_tempvars(self, cg: PyCodegen):
...
+            #
+            # For tensors without a source, we graph break for now.
+            #
+            # Handling the Handle: When a user retains the register_hook result in a handle, we intercept the
+            # STORE_FAST operation to record the user-designated local variable name. This ensures the reconstructed
+            # bytecode retains this name. If no handle is defined, we simply pop the generated value to keep the
+            # stack intact.
+            #
+            # Dynamo Tensor Hooks Workflow:
+            # - Functions passed to register_hook are lifted globally.
```
[the remark this answers is not in the document]
Comment (bdhirsh) [reply torch/_dynamo/side_effects.py:412]: I don't fully understand why we need to lift the hook into a global - why is it not sufficient to read it off of the source in some way? (maybe there's an existing autograd API to grab a hook from a tensor, or if not, maybe we can just stash it on a private attribute somewhere?)

File: torch/_dynamo/side_effects.py (hunk 3 above)
Comment (voznesenskym) [reply torch/_dynamo/side_effects.py:395]: The answer is even simpler :)

If you have an input, (or a param or global, but I lump those under inputs for this sake) - you have a source. The source is an effect, not a cause. The cause is that we know how to reference these objects in the user's invoking frame! Aka, they are a named or accessible object in f_locals/f_globals.

That means dynamo can, with relative ease, at the end of computing the graph for the forward, literally invoke:

`some_known_tensor.register_hook(...)` 

In the generated bytecode!

If you were to do that with an intermediary, you *still could* (in fact, an early prototype of this did that) but now you run afoul of preserving memory characteristics, because if you invoked `some_intermediary_tensor.register_hook(...)` in the generated bytecode, you have now forced `some_intermediary_tensor` to *not* be cleaned up / set to None, and have forced it to live alongside the rest of the graph AND get written out into outputs.

Seeing as the intermediary tensor is (a) fully encompassed within the graph (2) already has a clear lifecycle story - violating these things for the sake of hooks, and writing a bunch of messy code to preserve these structures is a no go.

File: torch/_dynamo/side_effects.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/side_effects.py:412]: This is a good point, let me see how far I get with a source.

The answer is sloppy, I initially had support for lambdas/locals (which would need to get lifted) but we don't need this anymore.

Give me a moment to change this

[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/side_effects.py:412]: ah we still need this to keep certain funcs around as references - but I have trimmed this down to reuse the original source where possible, and added a comment.

File: torch/_dynamo/variables/user_defined.py [hunk 5]
```diff
@@ -590,3 +590,33 @@ def __init__(self, value, **kwargs):
...
+            return [codegen.create_load_global(self.as_global, False, add=True)]
+        if self.last_seen_name:
+            # It is an invariant that at this point, a STORE_FAST was executed for this name.
+            return [codegen.create_load(self.last_seen_name)]
+        return super().reconstruct(codegen)
+
+    def rename(self, tx, name):
+        new_name = tx.output.new_var(name)
+        new_vt = self.clone(last_seen_name=new_name)
+        return tx.replace_all(self, new_vt)
```
Comment (ezyang) [inline torch/_dynamo/variables/user_defined.py:622]: I don't understand what's going on here

File: torch/_dynamo/variables/base.py [hunk 6]
```diff
@@ -297,6 +297,10 @@ def has_unpack_var_sequence(self, tx):
     def num_parameters(self):
         unimplemented(f"num_parameters: {self}")
 
+    def rename(self, tx, name):
+        # no-op
+        return self
```
Comment (ezyang) [inline torch/_dynamo/variables/base.py:302]: This needs a doc, this appears to be entirely new abstract method

File: torch/_dynamo/variables/user_defined.py (hunk 5 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/user_defined.py:622]: This is (1) declaring a new name 
(2) binding the name to a variable
(3) replacing all instances of the current variable with this named variable

File: torch/_dynamo/variables/base.py (hunk 6 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/base.py:302]: Yea, will document it for sure. It's kinda dependent on what we come up with w/r/t last_seen_name so I kept this minimal for now.

File: torch/_dynamo/variables/tensor.py [hunk 7]
```diff
@@ -644,6 +644,46 @@ def redistribute_fn_with_prim_types(x):
...
+                # This should not be onerous to support when needed.
+                unimplemented("NYI - lambda variables as hooks")
+            elif isinstance(fn_var, variables.functions.FunctoolsPartialVariable):
+                fn = fn_var.as_python_constant()
+                name = fn_var.func.fn.__name__
+            else:
+                fn = fn_var.fn
+                name = fn_var.fn.__name__
+
+            handle = self.as_proxy().node.meta["example_value"].register_hook(fn)
```
Comment (voznesenskym) [inline torch/_dynamo/variables/tensor.py:663]: tbd if we need this

Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:663]: Yea we do, its for handle.remove() subsumed by compile

**Comment by voznesenskym** [comment]
Discussed the handle side of things with @ezyang. Same page now.

File: torch/_dynamo/variables/user_defined.py [hunk 8]
```diff
@@ -590,3 +590,25 @@ def __init__(self, value, **kwargs):
...
+
+    # This reconstruct is actually pretty unique - it does not construct the object from scratch.
+    # Handles always come from a register_hook call on a tensor, and so, rerunning that for the codegen of a
+    # hook would be incorrect.
+    # Instead, the invariant is that codegen has already produced the handle and stored it at a known name.
+    def reconstruct(self, codegen):
+        if self.user_code_variable_name:
+            # It is an invariant that at this point, a STORE_FAST was executed for this name.
+            return [codegen.create_load(self.user_code_variable_name)]
+        return super().reconstruct(codegen)
```
Comment (ezyang) [inline torch/_dynamo/variables/user_defined.py:614]: This super... doesn't actually ever succeed right lol

File: torch/_dynamo/variables/tensor.py [hunk 9]
```diff
@@ -644,6 +644,46 @@ def redistribute_fn_with_prim_types(x):
...
+            fn_var = args[0]
+
+            if isinstance(fn_var, variables.NestedUserFunctionVariable):
+                # NestedUserFunctionVariable don't carry their fn, but reconstruction builds it
+                # This should not be onerous to support when needed.
+                unimplemented("NYI - lambda variables as hooks")
+            elif isinstance(fn_var, variables.functions.FunctoolsPartialVariable):
+                fn = fn_var.as_python_constant()
+                name = fn_var.func.fn.__name__
+            else:
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:659]: so... what kind of variable is fn_var here? 🤔

File: torch/_dynamo/variables/tensor.py (hunk 7 above)
Comment (ezyang) [reply torch/_dynamo/variables/tensor.py:663]: No, I don't understand. Why do you need the handle?

File: torch/_dynamo/variables/tensor.py [hunk 10]
```diff
@@ -644,6 +644,46 @@ def redistribute_fn_with_prim_types(x):
...
+                mutable_local=variables.base.MutableLocal(),
+                **options,
+            )
+            # Note - if the hook has no source, we are forced to lift it up into a global
+            # An example of where this might happen will be if the user creates a custom hook
+            # within a compile region, or wraps an func with a source (like a global)
+            # with a functools partial.  In that case, we can end up in a situation where we
+            # have a source for the internal func, but not the partial. The real fix here is to
+            # not always lift up to globals, but to use the lowest level scope possible to match the
+            # func lifecycle.
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:677]: No I don't understand. There's no inner scopes left after you compile. Don't you just need to store it long enough so that you can codegen a register hook for it? Then you can dump it, no global needed.

File: torch/_dynamo/variables/user_defined.py (hunk 8 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/user_defined.py:614]: yes, but the pattern is to fail out by propagating up. Makes it easier to compose stuff.

File: torch/_dynamo/variables/tensor.py (hunk 9 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:659]: Any callable - of which we have many.

File: torch/_dynamo/variables/tensor.py (hunk 10 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:677]: Thats true for inputs, but for intermediaries, we need to keep it alive. This lets us reuse that code. What we can do is lift a weakref tho - that will work.

Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:677]: I will change it to also only lift for intermeds

File: torch/_dynamo/variables/tensor.py (hunk 9 above)
Comment (ezyang) [reply torch/_dynamo/variables/tensor.py:659]: yes but why is func guaranteed to exist? It's not right?

File: torch/_dynamo/variables/tensor.py (hunk 7 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py]: Nah, I was wrong!!

File: torch/_dynamo/variables/tensor.py (hunk 9 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:659]: Sorry, which side of the else are you talking about?

File: torch/_dynamo/variables/tensor.py (hunk 10 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:675]: Actually, I dont need this.

File: torch/_dynamo/side_effects.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/side_effects.py:418]: I lied, we dont need this actually.

File: torch/_dynamo/variables/user_defined.py [hunk 11]
```diff
@@ -590,3 +590,31 @@ def __init__(self, value, **kwargs):
         super().__init__(value, **kwargs)
 
     # TODO Handle getattr for _length_per_key and _offset_per_key properly.
+
+
+class RemovableHandleVariable(VariableTracker):
+    def __init__(
+        self,
+        mutable_local=None,
+        idx=None,
```
Comment (ezyang) [inline torch/_dynamo/variables/user_defined.py:599]: would be nice to know what idx is doc here

File: torch/_dynamo/variables/tensor.py (hunk 9 above)
Comment (ezyang) [reply torch/_dynamo/variables/tensor.py:659]: err, fn, not func. Below here.

File: torch/_dynamo/variables/tensor.py [hunk 12]
```diff
@@ -644,6 +644,38 @@ def redistribute_fn_with_prim_types(x):
...
+                # This should not be onerous to support when needed.
+                unimplemented("NYI - lambda variables as hooks")
+            elif isinstance(fn_var, variables.functions.FunctoolsPartialVariable):
+                fn = fn_var.as_python_constant()
+                name = fn_var.func.fn.__name__
+            else:
+                fn = fn_var.fn
+                name = fn_var.fn.__name__
+
+            self.as_proxy().node.meta["example_value"].register_hook(fn)
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:663]: Do you need to even call this on the example value? Seems unnecessary.

File: torch/_dynamo/variables/functions.py [hunk 13]
```diff
@@ -604,13 +604,14 @@ def call_function(
...
 class FunctoolsPartialVariable(VariableTracker):
-    def __init__(self, func, args, keywords, **kwargs):
+    def __init__(self, func, args, keywords, original=None, **kwargs):
         super().__init__(**kwargs)
         self.func = func
         assert isinstance(args, list)
         self.args = args
         assert isinstance(keywords, dict)
         self.keywords = keywords
+        self.original = original
```
Comment (ezyang) [inline torch/_dynamo/variables/functions.py:614]: 🤔 This is weird. If original is set, I assume this is because you wrapped a functools.partial? But then how do you know the args/keywords??

Comment (ezyang) [reply torch/_dynamo/variables/functions.py:614]: oh because you read them out of the fields 🤔

File: torch/_dynamo/side_effects.py [hunk 14]
```diff
@@ -375,6 +380,77 @@ def codegen_save_tempvars(self, cg: PyCodegen):
...
             )
 
+    def register_hook(self, tensor, hook, handle):
+        idx = len(self.tensor_hooks)
+        self.tensor_hooks.append((tensor, hook, handle))
+        assert not handle.idx
+        handle.idx = idx
+
+    def remove_hook(self, idx):
+        del self.tensor_hooks[idx]
```
Comment (ezyang) [inline torch/_dynamo/side_effects.py:390]: 🚨  quadratic police 🚨

**Review by ezyang — APPROVED** [review]

File: torch/_dynamo/variables/functions.py (hunk 13 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/functions.py:614]: This is it, so it should be self contained.

File: torch/_dynamo/variables/tensor.py (hunk 9 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:662]: Its an invariant that anything that gets to here has a func, but we can graph break if its a surprising type
